hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java

```diff
@@ -66,168 +67,169 @@ private static ClientConfig newClientConfig() {
-                verify(listener, atLeastOnce()).attemptingToConnectToAddress(new Address("localhost", 5701));
-                verify(listener, atLeastOnce()).remoteClosedConnection(new Address("localhost", 5701));
-                verify(listener).clusterConnectionFailed(getTestMethodName());
-                verifyNoMoreInteractions(listener);
-            });
-        } finally {
-            do {
-                server.close();
-            } while (!server.isClosed());
+                    .getNetworkConfig()
+                    .addAddress(address.getHost() + ":" + address.getPort());
+
+            try {
+                HazelcastClient.newHazelcastClient(clientConfig);
+                fail("unexpectedly successful client startup");
+            } catch (IllegalStateException ignored) {
+                // Events are fired in a separate executor, it might take a while for
+                // the listeners to be notified.
+                assertTrueEventually(() -> {
+                    verify(listener, atLeastOnce()).possibleAddressesCollected(singletonList(address));
+                    verify(listener, atLeastOnce()).attemptingToConnectToAddress(address);
+                    verify(listener, atLeastOnce()).remoteClosedConnection(address);
+                    verify(listener).clusterConnectionFailed(getTestMethodName());
+                    verifyNoMoreInteractions(listener);
+                });
+            }
         }
     }
 
     @Test
-    public void clientNotAllowedInCluster() throws Exception {
-        String clusterName = getTestMethodName();
-        ClientConnectionProcessListener listener = mock(ClientConnectionProcessListener.class);
-
-        ClientConfig clientConfig = newClientConfig()
-                .addListenerConfig(new ListenerConfig(listener));
-        ClientFailoverConfig failoverConfig = new ClientFailoverConfig().setTryCount(0).addClientConfig(clientConfig);
+    public void clientNotAllowedInCluster() {
+        Config config = new Config();
+        config.setClusterName(getTestMethodName());
+        HazelcastInstanceImpl instance = getHazelcastInstanceImpl(Hazelcast.newHazelcastInstance(config));
+        instance.node.getClientEngine().applySelector(ClientSelectors.none());
+        Address address = instance.getCluster().getLocalMember().getAddress();
 
-        Hazelcast.newHazelcastInstance(new Config().setClusterName(clusterName));
+        ClientConnectionProcessListener listener = mock(ClientConnectionProcessListener.class);
+        ClientConfig clientConfig = newClientConfig();
+        clientConfig.addListenerConfig(new ListenerConfig(listener))
+                .getNetworkConfig()
+                .addAddress(address.getHost() + ":" + address.getPort());
 
         try {
-            HazelcastClient.newHazelcastFailoverClient(failoverConfig);
+            HazelcastClient.newHazelcastClient(clientConfig);
```

Comment:
Why did you change it to normal client?

Reply:
To make it simpler. The previous version of the test relied on some implementation details about when we threw that particular exception. With client selectors, it is more intuitive